src/gui/qgsmapcanvas.cpp

```diff
@@ -1595,6 +1613,25 @@ void QgsMapCanvas::setLayerStyleOverrides( const QMap<QString, QString> &overrid
   emit layerStyleOverridesChanged();
 }
 
+void QgsMapCanvas::setTheme( const QString &theme )
+{
+  if ( mTheme == theme )
+    return;
+
+  clearCache();
+  if ( theme.isEmpty() || !QgsProject::instance()->mapThemeCollection()->hasMapTheme( theme ) )
+  {
+    mTheme.clear();
+    mSettings.setLayerStyleOverrides( QMap< QString, QString>() );
+    emit themeChanged( QString() );
+  }
+  else
+  {
+    mTheme = theme;
+    setLayers( QgsProject::instance()->mapThemeCollection()->mapThemeVisibleLayers( mTheme ) );
+    emit themeChanged( theme );
+  }
```

Comment:
Would it be possible to add a comment how does it all work - i.e. why we are not changing style overrides and why are they fetched from theme collection in refreshMap() ? (it's a bit non-trivial to understand why things are like that)